<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from panda_arm.urdf.xacro           | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="panda">
  <link name="world"/>
  <!-- safety_distance: Minimum safety distance in [m] by which the collision volumes are expanded and which is enforced during robot motions -->
  <!-- arm_id: Namespace of the panda arm. Serves to differentiate between arms in case of multiple instances. -->
  <joint name="panda_joint_world" type="fixed">
    <parent link="world"/>
    <child link="panda_link0"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
  </joint>
  <!--<joint name="robot_to_world" type="fixed">
      <parent link="world" />
      <child link="${arm_id}_link0" />
      <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
      <dynamics damping="${joint_damping}" />
    </joint>-->
  <link name="panda_link0">
    <visual>
      <geometry>
        <mesh filename="/home/athena/franka_ros2_ws/install/franka_description/share/franka_description/meshes/visual/link0.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 1.5707963267948966 0" xyz="-0.075 0 0.06"/>
      <geometry>
        <cylinder length="0.03" radius="0.09"/>
      </geometry>
    </collision>
    <!-- <collision>
        <origin xyz="-0.06 0 0.06" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.06+safety_distance}" />
        </geometry>
      </collision>
      <collision>
        <origin xyz="-0.09 0 0.06" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.06+safety_distance}" />
        </geometry>
      </collision>-->
    <inertial>
      <origin rpy="0 0 0" xyz="-0.029190 -0.000879 0.056603"/>
      <mass value="3.06357"/>
      <inertia ixx="0.009" ixy="0.0" ixz="-0.00010" iyy="0.0122" iyz="0.0" izz="0.0122"/>
    </inertial>
  </link>
  <link name="panda_link1">
    <visual>
      <geometry>
        <mesh filename="/home/athena/franka_ros2_ws/install/franka_description/share/franka_description/meshes/visual/link1.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 -0.1915"/>
      <geometry>
        <cylinder length="0.2830" radius="0.09"/>
      </geometry>
    </collision>
    <!--<collision>
        <origin xyz="0 0 -0.333" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.06+safety_distance}" />
        </geometry>
      </collision>
      <collision>
        <origin xyz="0 0 -0.05" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.06+safety_distance}" />
        </geometry>
      </collision>-->
    <inertial>
      <origin rpy="0 0 0" xyz="0.000011 -0.031266 -0.069373"/>
      <mass value="2.34471"/>
      <inertia ixx="0.01418" ixy="0.0" ixz="0.0" iyy="0.01181" iyz="0.00236" izz="0.005515"/>
    </inertial>
  </link>
  <joint name="panda_joint1" type="revolute">
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973"/>
    <origin rpy="0 0 0" xyz="0 0 0.333"/>
    <parent link="panda_link0"/>
    <child link="panda_link1"/>
    <axis xyz="0 0 1"/>
    <limit effort="87" lower="-2.8973" upper="2.8973" velocity="2.1750"/>
    <dynamics damping="10.0"/>
  </joint>
  <link name="panda_link2">
    <visual>
      <geometry>
        <mesh filename="/home/athena/franka_ros2_ws/install/franka_description/share/franka_description/meshes/visual/link2.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.12" radius="0.09"/>
      </geometry>
    </collision>
    <!--<collision>
        <origin xyz="0 0 -0.06" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.06+safety_distance}" />
        </geometry>
      </collision>
      <collision>
        <origin xyz="0 0 0.06" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.06+safety_distance}" />
        </geometry>
      </collision> -->
    <inertial>
      <origin rpy="0 0 0" xyz="-0.000013 -0.070322 0.031178"/>
      <mass value="2.36414"/>
      <inertia ixx="0.014" ixy="0.0" ixz="0.0" iyy="0.0055" iyz="-0.0023" izz="0.01259"/>
    </inertial>
  </link>
  <joint name="panda_joint2" type="revolute">
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-1.7628" soft_upper_limit="1.7628"/>
    <origin rpy="-1.5707963267948966 0 0" xyz="0 0 0"/>
    <parent link="panda_link1"/>
    <child link="panda_link2"/>
    <axis xyz="0 0 1"/>
    <limit effort="87" lower="-1.7628" upper="1.7628" velocity="2.1750"/>
    <dynamics damping="10.0"/>
  </joint>
  <link name="panda_link3">
    <visual>
      <geometry>
        <mesh filename="/home/athena/franka_ros2_ws/install/franka_description/share/franka_description/meshes/visual/link3.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 -0.145"/>
      <geometry>
        <cylinder length="0.15" radius="0.09"/>
      </geometry>
    </collision>
    <!--      <collision>
        <origin xyz="0 0 -0.22" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.06+safety_distance}" />
        </geometry>
      </collision>
      <collision>
        <origin xyz="0 0 -0.07" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.06+safety_distance}" />
        </geometry>
      </collision>-->
    <inertial>
      <origin rpy="0 0 0" xyz="0.044348 0.024928 -0.038130"/>
      <mass value="2.3805"/>
      <inertia ixx="0.00716" ixy="-0.00102" ixz="-0.02040" iyy="0.00920" iyz="-0.00102" izz="0.007159"/>
    </inertial>
  </link>
  <joint name="panda_joint3" type="revolute">
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973"/>
    <origin rpy="1.5707963267948966 0 0" xyz="0 -0.316 0"/>
    <parent link="panda_link2"/>
    <child link="panda_link3"/>
    <axis xyz="0 0 1"/>
    <limit effort="87" lower="-2.8973" upper="2.8973" velocity="2.1750"/>
    <dynamics damping="10.0"/>
  </joint>
  <link name="panda_link4">
    <visual>
      <geometry>
        <mesh filename="/home/athena/franka_ros2_ws/install/franka_description/share/franka_description/meshes/visual/link4.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.12" radius="0.09"/>
      </geometry>
    </collision>
    <!--<collision>
        <origin xyz="0 0 0.06" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.06+safety_distance}" />
        </geometry>
      </collision>
      <collision>
        <origin xyz="0 0 -0.06" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.06+safety_distance}" />
        </geometry>
      </collision>-->
    <inertial>
      <origin rpy="0 0 0" xyz="-0.038550 0.039526 0.024716"/>
      <mass value="2.42754"/>
      <inertia ixx="0.008164" ixy="0.003061" ixz="0.00102" iyy="0.008165" iyz="0.00102" izz="0.009185"/>
    </inertial>
  </link>
  <joint name="panda_joint4" type="revolute">
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-3.0718" soft_upper_limit="-0.0698"/>
    <origin rpy="1.5707963267948966 0 0" xyz="0.0825 0 0"/>
    <parent link="panda_link3"/>
    <child link="panda_link4"/>
    <axis xyz="0 0 1"/>
    <limit effort="87" lower="-3.0718" upper="-0.0698" velocity="2.1750"/>
    <dynamics damping="10.0"/>
  </joint>
  <link name="panda_link5">
    <visual>
      <geometry>
        <mesh filename="/home/athena/franka_ros2_ws/install/franka_description/share/franka_description/meshes/visual/link5.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 -0.26"/>
      <geometry>
        <cylinder length="0.1" radius="0.09"/>
      </geometry>
    </collision>
    <!-- <collision>
        <origin xyz="0 0 -0.31" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.06+safety_distance}" />
        </geometry>
      </collision>
      <collision>
        <origin xyz="0 0 -0.21" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.06+safety_distance}" />
        </geometry>
      </collision>
      <collision>
        <origin xyz="0 0.08 -0.13" rpy="0 0 0"/>
        <geometry>
          <cylinder radius="${0.025+safety_distance}" length="0.14" />
        </geometry>
      </collision>
      <collision>
        <origin xyz="0 0.08 -0.06" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.025+safety_distance}" />
        </geometry>
      </collision>
      <collision>
        <origin xyz="0 0.08 -0.20" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.025+safety_distance}" />
        </geometry>
      </collision>-->
    <inertial>
      <origin rpy="0 0 0" xyz="-0.000064 0.038412 -0.109970"/>
      <mass value="3.49611"/>
      <inertia ixx="0.0307" ixy="0.0" ixz="0.0" iyy="0.02761" iyz="-0.0613" izz="0.00818"/>
    </inertial>
  </link>
  <joint name="panda_joint5" type="revolute">
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973"/>
    <origin rpy="-1.5707963267948966 0 0" xyz="-0.0825 0.384 0"/>
    <parent link="panda_link4"/>
    <child link="panda_link5"/>
    <axis xyz="0 0 1"/>
    <limit effort="12" lower="-2.8973" upper="2.8973" velocity="2.6100"/>
    <dynamics damping="10.0"/>
  </joint>
  <link name="panda_link6">
    <visual>
      <geometry>
        <mesh filename="/home/athena/franka_ros2_ws/install/franka_description/share/franka_description/meshes/visual/link6.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 -0.03"/>
      <geometry>
        <cylinder length="0.08" radius="0.08"/>
      </geometry>
    </collision>
    <!--<collision>
        <origin xyz="0 0 0.01" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.05+safety_distance}" />
        </geometry>
      </collision>
      <collision>
        <origin xyz="0 0 -0.07" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.05+safety_distance}" />
        </geometry>
      </collision> -->
    <inertial>
      <origin rpy="0 0 0" xyz="0.051002 0.006933 0.006169"/>
      <mass value="1.46736"/>
      <inertia ixx="0.00205" ixy="0.0" ixz="0.0" iyy="0.004089" iyz="0" izz="0.00490"/>
    </inertial>
  </link>
  <joint name="panda_joint6" type="revolute">
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-0.0175" soft_upper_limit="3.7525"/>
    <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
    <parent link="panda_link5"/>
    <child link="panda_link6"/>
    <axis xyz="0 0 1"/>
    <limit effort="12" lower="-0.0175" upper="3.7525" velocity="2.6100"/>
    <dynamics damping="10.0"/>
  </joint>
  <link name="panda_link7">
    <visual>
      <geometry>
        <mesh filename="/home/athena/franka_ros2_ws/install/franka_description/share/franka_description/meshes/visual/link7.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0.01"/>
      <geometry>
        <cylinder length="0.14" radius="0.07"/>
      </geometry>
    </collision>
    <!--<collision>
        <origin xyz="0 0 0.08" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.04+safety_distance}"  />
        </geometry>
      </collision>
      <collision>
        <origin xyz="0 0 -0.06" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.04+safety_distance}"  />
        </geometry>
      </collision>-->
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.45606"/>
      <inertia ixx="0.3" ixy="0.0" ixz="0.0" iyy="0.3" iyz="0.0" izz="0.3"/>
    </inertial>
  </link>
  <joint name="panda_joint7" type="revolute">
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973"/>
    <origin rpy="1.5707963267948966 0 0" xyz="0.088 0 0"/>
    <parent link="panda_link6"/>
    <child link="panda_link7"/>
    <axis xyz="0 0 1"/>
    <limit effort="12" lower="-2.8973" upper="2.8973" velocity="2.6100"/>
    <dynamics damping="10.0"/>
  </joint>
  <link name="panda_link8">
    <collision>
      <origin rpy="3.141592653589793 1.5707963267948966 1.5707963267948966" xyz="0.0424 0.0424 -0.0250"/>
      <geometry>
        <cylinder length="0.01" radius="0.06"/>
      </geometry>
    </collision>
    <!--<collision>
        <origin xyz="0.0424 0.0424 -0.02" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.03+safety_distance}"  />
        </geometry>
      </collision>
      <collision>
        <origin xyz="0.0424 0.0424 -0.03" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.03+safety_distance}"  />
        </geometry>
      </collision>-->
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.001"/>
      <inertia ixx="0.3" ixy="0.0" ixz="0.0" iyy="0.3" iyz="0.0" izz="0.3"/>
    </inertial>
  </link>
  <joint name="panda_joint8" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0.107"/>
    <parent link="panda_link7"/>
    <child link="panda_link8"/>
  </joint>
  <joint name="panda_hand_joint" type="fixed">
    <parent link="panda_link8"/>
    <child link="panda_hand"/>
    <origin rpy="0 0 -0.7853981633974483" xyz="0 0 0"/>
  </joint>
  <link name="panda_hand">
    <visual>
      <geometry>
        <mesh filename="/home/athena/franka_ros2_ws/install/franka_description/share/franka_description/meshes/visual/hand.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 1.5707963267948966 1.5707963267948966" xyz="0 0 0.04"/>
      <geometry>
        <cylinder length="0.1" radius="0.07"/>
      </geometry>
    </collision>
    <!--<collision>
        <origin xyz="0 -0.05 0.04" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.04+safety_distance}"/>
        </geometry>
      </collision>
      <collision>
        <origin xyz="0 0.05 0.04" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.04+safety_distance}"/>
        </geometry>
      </collision>
      <collision>
        <origin xyz="0 0 0.1" rpy="0 ${pi/2} ${pi/2}"/>
        <geometry>
          <cylinder radius="${0.02+safety_distance}" length="0.1"/>
        </geometry>
      </collision>
      <collision>
        <origin xyz="0 -0.05 0.1" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.02+safety_distance}"/>
        </geometry>
      </collision>
      <collision>
        <origin xyz="0 0.05 0.1" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${0.02+safety_distance}"/>
        </geometry>
      </collision> -->
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.68"/>
      <inertia ixx="0.1" ixy="0.0" ixz="0.0" iyy="0.1" iyz="0.0" izz="0.1"/>
    </inertial>
  </link>
  <!-- Define the hand_tcp frame -->
  <link name="panda_hand_tcp"/>
  <joint name="panda_hand_tcp_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0.1034"/>
    <parent link="panda_hand"/>
    <child link="panda_hand_tcp"/>
  </joint>
  <link name="panda_leftfinger">
    <visual>
      <origin rpy="0 0 -3.141592653589793" xyz="0 0 0"/>
      <geometry>
        <mesh filename="/home/athena/franka_ros2_ws/install/franka_description/share/franka_description/meshes/visual/finger.dae"/>
      </geometry>
    </visual>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.02"/>
      <inertia ixx="0.000001667" ixy="0.0" ixz="0.0" iyy="0.000001667" iyz="0.0" izz="0.000000667"/>
    </inertial>
  </link>
  <link name="panda_rightfinger">
    <visual>
      <origin rpy="0 0 3.141592653589793" xyz="0 0 0"/>
      <geometry>
        <mesh filename="/home/athena/franka_ros2_ws/install/franka_description/share/franka_description/meshes/visual/finger.dae"/>
      </geometry>
    </visual>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.02"/>
      <inertia ixx="0.000001667" ixy="0.0" ixz="0.0" iyy="0.000001667" iyz="0.0" izz="0.000000667"/>
    </inertial>
  </link>
  <joint name="panda_finger_joint1" type="prismatic">
    <parent link="panda_hand"/>
    <child link="panda_leftfinger"/>
    <origin rpy="0 0 0" xyz="0 0 0.0584"/>
    <axis xyz="0 1 0"/>
    <limit effort="20" lower="0.0" upper="0.04" velocity="0.2"/>
  </joint>
  <joint name="panda_finger_joint2" type="prismatic">
    <parent link="panda_hand"/>
    <child link="panda_rightfinger"/>
    <origin rpy="0 0 0" xyz="0 0 0.0584"/>
    <axis xyz="0 -1 0"/>
    <limit effort="20" lower="0.0" upper="0.04" velocity="0.2"/>
    <mimic joint="panda_finger_joint1"/>
  </joint>
  <gazebo>
    <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
      <parameters>/home/athena/franka_ros2_ws/install/franka_description/share/franka_description/config/controllers.yaml</parameters>
    </plugin>
  </gazebo>
  <ros2_control name="GazeboSystem" type="system">
    <hardware>
      <plugin>gazebo_ros2_control/GazeboSystem</plugin>
    </hardware>
    <joint name="panda_joint1">
      <param name="initial_position">0.0</param>
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="panda_joint2">
      <param name="initial_position">-0.7853981633974483</param>
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="panda_joint3">
      <param name="initial_position">0.0</param>
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="panda_joint4">
      <param name="initial_position">-2.356194490192345</param>
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="panda_joint5">
      <param name="initial_position">0.0</param>
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="panda_joint6">
      <param name="initial_position">1.5707963267948966</param>
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="panda_joint7">
      <param name="initial_position">0.7853981633974483</param>
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
  </ros2_control>
  <!-- publishing the joint states to the /JointState topic-->
  <gazebo>
    <plugin filename="libgazebo_ros_joint_state_publisher.so" name="gazebo_ros_joint_state_publisher">
      <update_rate>1000</update_rate>
      <joint_name>panda_joint1</joint_name>
      <joint_name>panda_joint2</joint_name>
      <joint_name>panda_joint3</joint_name>
      <joint_name>panda_joint4</joint_name>
      <joint_name>panda_joint5</joint_name>
      <joint_name>panda_joint6</joint_name>
      <joint_name>panda_joint7</joint_name>
    </plugin>
  </gazebo>
  <!-- Friction parameters-->
  <gazebo reference="world">
    <material>Gazebo/Grey</material>
    <mu1>2</mu1>
    <mu2>2</mu2>
  </gazebo>
  <gazebo reference="panda_link0">
    <kp>0.0</kp>
    <kd>0.0</kd>
    <min_depth>0.001</min_depth>
    <!-- <max_vel>0</max_vel> -->
    <slip1>0</slip1>
    <slip2>0</slip2>
    <mu1>0</mu1>
    <mu2>0</mu2>
  </gazebo>
  <!-- Link1 -->
  <gazebo reference="panda_link1">
    <kp>0.0</kp>
    <kd>0.0</kd>
    <min_depth>0.001</min_depth>
    <!-- <max_vel>0</max_vel> -->
    <slip1>0</slip1>
    <slip2>0</slip2>
    <mu1>0</mu1>
    <mu2>0</mu2>
  </gazebo>
  <!-- Link2 -->
  <gazebo reference="panda_link2">
    <mu1>0.61</mu1>
    <mu2>0.61</mu2>
  </gazebo>
  <!-- Link3 -->
  <gazebo reference="panda_link3">
    <mu1>0.61</mu1>
    <mu2>0.61</mu2>
  </gazebo>
  <!-- Link4 -->
  <gazebo reference="panda_link4">
    <mu1>0.61</mu1>
    <mu2>0.61</mu2>
  </gazebo>
  <!-- Link5 -->
  <gazebo reference="panda_link5">
    <mu1>0.61</mu1>
    <mu2>0.61</mu2>
  </gazebo>
  <!-- Link6 -->
  <gazebo reference="panda_link6">
    <kp>100000.0</kp>
    <kd>100.0</kd>
    <min_depth>0.001</min_depth>
    <!-- <max_vel>0</max_vel> -->
    <slip1>0.2</slip1>
    <slip2>0.2</slip2>
    <mu>0.81</mu>
    <mu2>0.5</mu2>
  </gazebo>
  <gazebo reference="panda_link7">
    <!-- <kp>50000.0</kp> -->
    <kp>50000.0</kp>
    <kd>1.0</kd>
    <min_depth>0.001</min_depth>
    <!-- <max_vel>0</max_vel> -->
    <slip1>0</slip1>
    <slip2>0</slip2>
    <mu1>0</mu1>
    <mu2>0</mu2>
  </gazebo>
</robot>
